File: platform/polyglot/rag_ui_component_ingest_config/py/rag_ui_component_ingest_config/base_config.py
"""BaseIngestConfig — shared infrastructure configuration resolved from env vars."""

from __future__ import annotations

import os
from dataclasses import dataclass, field

from env_resolver import resolve_anthropic_env
from .defaults import DEFAULTS
from .logger import create_logger
from .schema import BaseIngestConfigSchema

_log = create_logger("base_config")


def _env_str(name: str, default: str) -> str:
    return os.environ.get(name, default)


def _env_int(name: str, default: int) -> int:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        return int(raw)
    except ValueError:
        _log.warning("Invalid integer for %s=%r, using default %d", name, raw, default)
        return default


def _env_float(name: str, default: float) -> float:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        return float(raw)
    except ValueError:
        _log.warning("Invalid float for %s=%r, using default %f", name, raw, default)
        return default


def _env_bool(name: str, default: bool) -> bool:
    raw = os.environ.get(name)
    if raw is None:
        return default
    return raw.strip().lower() in ("1", "true", "yes", "on")


@dataclass(frozen=True)
class BaseIngestConfig:
    """Resolved shared infrastructure configuration.

    All fields are populated from environment variables with fallback to the
    values defined in ``defaults.DEFAULTS``. Validation is performed via
    ``BaseIngestConfigSchema`` in ``__post_init__``.
    """

    # Paths
    dataset_root: str = field(
        default_factory=lambda: _env_str("DATASET_ROOT", DEFAULTS["dataset_root"])
    )
    persist_root: str = field(
        default_factory=lambda: _env_str("RAG_PERSIST_ROOT", DEFAULTS["persist_root"])
    )

    # Embeddings
    embeddings_model_name: str = field(
        default_factory=lambda: _env_str(
            "EMBEDDINGS_MODEL_NAME", DEFAULTS["embeddings_model_name"]
        )
    )

    # Chunking
    chunk_size: int = field(
        default_factory=lambda: _env_int("CHUNK_SIZE", DEFAULTS["chunk_size"])
    )
    chunk_overlap: int = field(
        default_factory=lambda: _env_int("CHUNK_OVERLAP", DEFAULTS["chunk_overlap"])
    )

    # Vector backend
    vector_backend: str = field(
        default_factory=lambda: _env_str("RAG_VECTOR_BACKEND", DEFAULTS["vector_backend"])
    )

    # Elasticsearch
    elasticsearch_host: str = field(
        default_factory=lambda: _env_str("ELASTIC_DB_HOST", DEFAULTS["elasticsearch_host"])
    )
    elasticsearch_port: int = field(
        default_factory=lambda: _env_int("ELASTIC_DB_PORT", DEFAULTS["elasticsearch_port"])
    )
    elasticsearch_scheme: str = field(
        default_factory=lambda: _env_str("ELASTIC_DB_SCHEME", DEFAULTS["elasticsearch_scheme"])
    )

    # Redis
    redis_host: str = field(
        default_factory=lambda: _env_str("REDIS_HOST", DEFAULTS["redis_host"])
    )
    redis_port: int = field(
        default_factory=lambda: _env_int("REDIS_PORT", DEFAULTS["redis_port"])
    )

    # LLM
    llm_provider: str = field(
        default_factory=lambda: _env_str("LLM_PROVIDER", DEFAULTS["llm_provider"])
    )
    openai_model: str = field(
        default_factory=lambda: _env_str("OPENAI_MODEL", DEFAULTS["openai_model"])
    )
    anthropic_model: str = field(
        default_factory=lambda: resolve_anthropic_env().model
    )
    gemini_model: str = field(
        default_factory=lambda: _env_str("GEMINI_MODEL", DEFAULTS["gemini_model"])
    )

    # Search / retrieval
    hybrid_alpha: float = field(
        default_factory=lambda: _env_float("HYBRID_ALPHA", DEFAULTS["hybrid_alpha"])
    )
    score_threshold: float = field(
        default_factory=lambda: _env_float(
            "SCORE_THRESHOLD", DEFAULTS["score_threshold"]
        )
    )
    reranker_enabled: bool = field(
        default_factory=lambda: _env_bool(
            "RERANKER_ENABLED", DEFAULTS["reranker_enabled"]
        )
    )
    reranker_model: str = field(
        default_factory=lambda: _env_str("RERANKER_MODEL", DEFAULTS["reranker_model"])
    )
    retrieve_n: int = field(
        default_factory=lambda: _env_int("RETRIEVE_N", DEFAULTS["retrieve_n"])
    )
    top_k: int = field(
        default_factory=lambda: _env_int("TOP_K", DEFAULTS["top_k"])
    )

    # Database
    postgres_enabled: bool = field(
        default_factory=lambda: _env_bool(
            "RAG_POSTGRES_ENABLED", DEFAULTS["postgres_enabled"]
        )
    )

    # ------------------------------------------------------------------
    # Post-init validation
    # ------------------------------------------------------------------

    def __post_init__(self) -> None:
        schema = BaseIngestConfigSchema(
            dataset_root=self.dataset_root,
            persist_root=self.persist_root,
            embeddings_model_name=self.embeddings_model_name,
            chunk_size=self.chunk_size,
            chunk_overlap=self.chunk_overlap,
            vector_backend=self.vector_backend,  # type: ignore[arg-type]
            elasticsearch_host=self.elasticsearch_host,
            elasticsearch_port=self.elasticsearch_port,
            elasticsearch_scheme=self.elasticsearch_scheme,
            redis_host=self.redis_host,
            redis_port=self.redis_port,
            llm_provider=self.llm_provider,
            openai_model=self.openai_model,
            anthropic_model=self.anthropic_model,
            gemini_model=self.gemini_model,
            hybrid_alpha=self.hybrid_alpha,
            score_threshold=self.score_threshold,
            reranker_enabled=self.reranker_enabled,
            reranker_model=self.reranker_model,
            retrieve_n=self.retrieve_n,
            top_k=self.top_k,
            postgres_enabled=self.postgres_enabled,
        )
        _log.debug(
            "BaseIngestConfig validated: %s",
            schema.model_dump(),
        )

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict[str, object]:
        """Return a snake_case dictionary of all resolved values."""
        return {
            "dataset_root": self.dataset_root,
            "persist_root": self.persist_root,
            "embeddings_model_name": self.embeddings_model_name,
            "chunk_size": self.chunk_size,
            "chunk_overlap": self.chunk_overlap,
            "vector_backend": self.vector_backend,
            "elasticsearch_host": self.elasticsearch_host,
            "elasticsearch_port": self.elasticsearch_port,
            "elasticsearch_scheme": self.elasticsearch_scheme,
            "redis_host": self.redis_host,
            "redis_port": self.redis_port,
            "llm_provider": self.llm_provider,
            "openai_model": self.openai_model,
            "anthropic_model": self.anthropic_model,
            "gemini_model": self.gemini_model,
            "hybrid_alpha": self.hybrid_alpha,
            "score_threshold": self.score_threshold,
            "reranker_enabled": self.reranker_enabled,
            "reranker_model": self.reranker_model,
            "retrieve_n": self.retrieve_n,
            "top_k": self.top_k,
            "postgres_enabled": self.postgres_enabled,
        }

File: platform/polyglot/rag_ui_component_ingest_config/mjs/src/base-config.mjs
/**
 * @fileoverview BaseIngestConfig — shared infrastructure configuration resolved
 * from environment variables with fallback to DEFAULTS.
 *
 * Immutable after construction (Object.freeze).
 * `toDict()` / `toJSON()` emit snake_case for wire-format compatibility with
 * the Python counterpart.
 */

import { DEFAULTS } from './defaults.mjs';
import { BaseIngestConfigSchema } from './schema.mjs';
import { createLogger } from './logger.mjs';
import { resolveAnthropicEnv } from '@internal/env-resolver';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Read an env var and return a string, or fall back to `defaultValue`.
 *
 * @param {string} name
 * @param {string|number|boolean} defaultValue
 * @returns {string|undefined}
 */
function env(name, defaultValue) {
  const raw = process.env[name];
  return raw !== undefined ? raw : defaultValue;
}

/**
 * Parse an env var as an integer, with a numeric fallback.
 *
 * @param {string} name
 * @param {number} defaultValue
 * @returns {number}
 */
function envInt(name, defaultValue) {
  const raw = process.env[name];
  if (raw !== undefined) {
    const parsed = parseInt(raw, 10);
    if (!Number.isNaN(parsed)) return parsed;
  }
  return defaultValue;
}

/**
 * Parse an env var as a float, with a numeric fallback.
 *
 * @param {string} name
 * @param {number} defaultValue
 * @returns {number}
 */
function envFloat(name, defaultValue) {
  const raw = process.env[name];
  if (raw !== undefined) {
    const parsed = parseFloat(raw);
    if (!Number.isNaN(parsed)) return parsed;
  }
  return defaultValue;
}

/**
 * Parse an env var as a boolean.
 * Truthy strings: "1", "true", "yes" (case-insensitive).
 *
 * @param {string} name
 * @param {boolean} defaultValue
 * @returns {boolean}
 */
function envBool(name, defaultValue) {
  const raw = process.env[name];
  if (raw !== undefined) {
    return ['1', 'true', 'yes'].includes(raw.toLowerCase());
  }
  return defaultValue;
}

// ---------------------------------------------------------------------------
// BaseIngestConfig
// ---------------------------------------------------------------------------

/**
 * Shared infrastructure configuration resolved from environment variables with
 * fallback to frozen `DEFAULTS`.
 *
 * Immutable after construction.  Wire serialisation (toDict / toJSON) uses
 * snake_case to match the Python counterpart and the REST/protobuf contract.
 */
export class BaseIngestConfig {
  /**
   * @param {object} [overrides={}] - Optional explicit overrides (camelCase).
   *   When an override is present it takes precedence over both the env var and
   *   the compiled-in default.
   * @param {object|null} [parentLogger=null] - Pino-compatible logger.
   */
  constructor(overrides = {}, parentLogger = null) {
    this._log = createLogger('BaseIngestConfig', parentLogger);

    // Resolve each field: override > env var > DEFAULTS
    const resolved = {
      datasetRoot:          overrides.datasetRoot          ?? env('DATASET_ROOT',           DEFAULTS.datasetRoot),
      persistRoot:          overrides.persistRoot          ?? env('RAG_PERSIST_ROOT',        DEFAULTS.persistRoot),
      embeddingsModelName:  overrides.embeddingsModelName  ?? env('EMBEDDINGS_MODEL_NAME',   DEFAULTS.embeddingsModelName),

      chunkSize:            overrides.chunkSize            ?? envInt('CHUNK_SIZE',           DEFAULTS.chunkSize),
      chunkOverlap:         overrides.chunkOverlap         ?? envInt('CHUNK_OVERLAP',        DEFAULTS.chunkOverlap),

      vectorBackend:        overrides.vectorBackend        ?? env('RAG_VECTOR_BACKEND',      DEFAULTS.vectorBackend),

      elasticsearchHost:    overrides.elasticsearchHost    ?? env('ELASTIC_DB_HOST',         DEFAULTS.elasticsearchHost),
      elasticsearchPort:    overrides.elasticsearchPort    ?? envInt('ELASTIC_DB_PORT',      DEFAULTS.elasticsearchPort),
      elasticsearchScheme:  overrides.elasticsearchScheme  ?? env('ELASTIC_DB_SCHEME',       DEFAULTS.elasticsearchScheme),

      redisHost:            overrides.redisHost            ?? env('REDIS_HOST',              DEFAULTS.redisHost),
      redisPort:            overrides.redisPort            ?? envInt('REDIS_PORT',           DEFAULTS.redisPort),

      llmProvider:          overrides.llmProvider          ?? env('LLM_PROVIDER',            DEFAULTS.llmProvider),
      openaiModel:          overrides.openaiModel          ?? env('OPENAI_MODEL',            DEFAULTS.openaiModel),
      anthropicModel:       overrides.anthropicModel       ?? resolveAnthropicEnv().model,
      geminiModel:          overrides.geminiModel          ?? env('GEMINI_MODEL',            DEFAULTS.geminiModel),

      hybridAlpha:          overrides.hybridAlpha          ?? envFloat('HYBRID_ALPHA',       DEFAULTS.hybridAlpha),
      scoreThreshold:       overrides.scoreThreshold       ?? envFloat('SCORE_THRESHOLD',    DEFAULTS.scoreThreshold),
      rerankerEnabled:      overrides.rerankerEnabled      ?? envBool('RERANKER_ENABLED',    DEFAULTS.rerankerEnabled),
      rerankerModel:        overrides.rerankerModel        ?? env('RERANKER_MODEL',          DEFAULTS.rerankerModel),
      retrieveN:            overrides.retrieveN            ?? envInt('RETRIEVE_N',           DEFAULTS.retrieveN),
      topK:                 overrides.topK                 ?? envInt('TOP_K',               DEFAULTS.topK),

      postgresEnabled:      overrides.postgresEnabled      ?? envBool('RAG_POSTGRES_ENABLED', DEFAULTS.postgresEnabled),
    };

    // Zod validation — throws ZodError on invalid input
    const parsed = BaseIngestConfigSchema.parse(resolved);

    // Assign all validated fields directly to `this`
    Object.assign(this, parsed);

    Object.freeze(this);

    this._log.debug('BaseIngestConfig constructed', {
      vectorBackend: this.vectorBackend,
      llmProvider: this.llmProvider,
    });
  }

  // ---------------------------------------------------------------------------
  // Serialisation
  // ---------------------------------------------------------------------------

  /**
   * Return a plain object with snake_case keys for wire-format compatibility.
   *
   * @returns {Record<string, unknown>}
   */
  toDict() {
    return {
      dataset_root:          this.datasetRoot,
      persist_root:          this.persistRoot,
      embeddings_model_name: this.embeddingsModelName,
      chunk_size:            this.chunkSize,
      chunk_overlap:         this.chunkOverlap,
      vector_backend:        this.vectorBackend,
      elasticsearch_host:    this.elasticsearchHost,
      elasticsearch_port:    this.elasticsearchPort,
      elasticsearch_scheme:  this.elasticsearchScheme,
      redis_host:            this.redisHost,
      redis_port:            this.redisPort,
      llm_provider:          this.llmProvider,
      openai_model:          this.openaiModel,
      anthropic_model:       this.anthropicModel,
      gemini_model:          this.geminiModel,
      hybrid_alpha:          this.hybridAlpha,
      score_threshold:       this.scoreThreshold,
      reranker_enabled:      this.rerankerEnabled,
      reranker_model:        this.rerankerModel,
      retrieve_n:            this.retrieveN,
      top_k:                 this.topK,
      postgres_enabled:      this.postgresEnabled,
    };
  }

  /**
   * Alias for `toDict()` — called by JSON.stringify automatically.
   *
   * @returns {Record<string, unknown>}
   */
  toJSON() {
    return this.toDict();
  }
}
